onecodex/viz/_custom_plots/pyodide.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+import json
+
+import requests
+
+from .collection import SampleCollection, Samples
+from .models import PlotParams
+from .enums import SuggestionType
+
+if TYPE_CHECKING:
+    from pyodide.ffi import JsProxy
+
+
+def plot(params: JsProxy, csrf_token: str, include_exported_chart_data: bool = False) -> dict:
+    import js  # available from pyodide
+
+    # TODO only configure theme a single time when pyodide loads
+    from onecodex.viz import configure_onecodex_theme
+
+    configure_onecodex_theme()
+
+    params = PlotParams.model_validate(_replace_jsnull(params.to_py()))
+    cache = globals().get("CUSTOM_PLOTS_CACHE", {})
+
+    uuid = None
+    type_ = None
+    if params.tag:
+        uuid = params.tag
+        type_ = SuggestionType.Tag
+    elif params.project:
+        uuid = params.project
+        type_ = SuggestionType.Project
+    else:
+        raise Exception
+
+    key = (uuid, params.metric)
+    if key in cache:
+        collection = cache[key]
+    else:
+        base_url = js.self.location.origin
+        url = f"{base_url}/api/v2/custom-plots/sample-data"
+
+        samples = []
+        page = 1
+        while True:
```

Comment:
Nit: this relies on sane responses from the API and could potentially be dangerous and block user's browser. Let's maybe introduce an arbitrary max value, e.g. 1000 or something?

Reply:
This will be run in a separate web worker thread to prevent blocking the user's browser. I'd rather keep this implementation because it matches how we implement "load all paginated results" in other parts of the codebase (e.g. `onecodex/models/base.py` line ~346)